### PR TITLE
Fixtypings

### PR DESCRIPTION
**Pull request by SanthoshBonala**
## Proposed changes

 
Fixes https://github.com/webdriverio/webdriverio/issues/4517

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/webdriverio/webdriverio.d.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ type ElementPromise = Omit<WebdriverIO.Element,
     | 'elementId'
     | 'element-6066-11e4-a52e-4f735466cecf'
     | 'ELEMENT'
+    | 'dragAndDrop'
```
Comment (mgrybyk) [inline packages/webdriverio/webdriverio.d.ts]: Why do we need to define commands here? They have to be declared once in webdriverio.core.d.ts and automatically wrapped with Promise here

Comment (SanthoshBonala) [reply packages/webdriverio/webdriverio.d.ts]: Here the problem is with arguments not return types. Element in dragAndDrop in webdriverio-core.d.ts refers to WebdriverIO.Element always but in async mode it needs to refer to WebdriverIOAsync.Element

**Comment by christian-bromann** [comment]
This also needs tests in https://github.com/webdriverio/webdriverio/tree/master/tests/typings

**Comment by SanthoshBonala** [comment]
Sure will be adding in a couple of days

**Comment by mgrybyk** [comment]
Please add tests to both sync and async modes
- https://github.com/webdriverio/webdriverio/blob/master/tests/typings/sync/sync.ts
- https://github.com/webdriverio/webdriverio/blob/master/tests/typings/webdriverio/async.ts

Thank you!

File: scripts/templates/webdriverio.tpl.d.ts [hunk 2]
```diff
@@ -147,14 +147,14 @@ declare namespace WebdriverIO {
     interface Hooks extends _HooksArray, _Hooks { }
 
     type ActionTypes = 'press' | 'longPress' | 'tap' | 'moveTo' | 'wait' | 'release';
-    interface TouchAction {
+    interface TouchAction<T> {
         action: ActionTypes,
         x?: number,
         y?: number,
-        element?: Element,
+        element?: T,
```
Comment (baruchvlz) [inline scripts/templates/webdriverio.tpl.d.ts]: Can a touch action be done in something else that is not an Element?

Comment (christian-bromann) [reply scripts/templates/webdriverio.tpl.d.ts]: yes, the property is optional so you can do an action on arbitrary coordinates

Comment (christian-bromann) [reply scripts/templates/webdriverio.tpl.d.ts]: @SanthoshBonala it seems that @baruchvlz is correct and this property only takes Element types and nothing else

File: tests/typings/webdriverio/async.ts [hunk 3]
```diff
@@ -114,4 +131,5 @@ const config: WebdriverIOAsync.Config = {
 // allure-reporter
 allure.addFeature('')
 
+
```
Comment (christian-bromann) [inline tests/typings/webdriverio/async.ts]:
```suggestion
```

File: tests/typings/webdriverio/async.ts [hunk 4]
```diff
@@ -103,6 +103,23 @@ async function bar() {
...
+        x: 0,
+        y: 0
+    }
+    ele.touchAction(touchAction)
+    browser.touchAction(touchAction)
+    
+    // dragAndDrop
+    ele.dragAndDrop(ele, 0)
+
+
```
Comment (christian-bromann) [inline tests/typings/webdriverio/async.ts]:
```suggestion
```

File: tests/typings/webdriverio/async.ts [hunk 5]
```diff
@@ -103,6 +103,23 @@ async function bar() {
...
+        ms: 0,
+        x: 0,
+        y: 0
+    }
+    ele.touchAction(touchAction)
+    browser.touchAction(touchAction)
+    
+    // dragAndDrop
+    ele.dragAndDrop(ele, 0)
+
```
Comment (christian-bromann) [inline tests/typings/webdriverio/async.ts]:
```suggestion
```

File: tests/typings/webdriverio/async.ts [hunk 6]
```diff
@@ -103,6 +103,23 @@ async function bar() {
...
+        element: await $(''),
+        ms: 0,
+        x: 0,
+        y: 0
+    }
+    ele.touchAction(touchAction)
+    browser.touchAction(touchAction)
+    
+    // dragAndDrop
+    ele.dragAndDrop(ele, 0)
```
Comment (christian-bromann) [inline tests/typings/webdriverio/async.ts]: shouldn't all the commands above also be called with `await`?

**Comment by christian-bromann** [comment]
@mgrybyk any objections?

**Comment by SanthoshBonala** [comment]
why the tests are failing? Is this a known issue?

**Comment by christian-bromann** [comment]
@SanthoshBonala can you update your branch? It should fix the problem.

**Comment by SanthoshBonala** [comment]
> @SanthoshBonala can you update your branch? It should fix the problem.

still fails. I don't think it's related to my change

**Comment by mgrybyk** [comment]
It's not your fault, you need to rebase with master.
Or we can just merge it as is

**Comment by SanthoshBonala** [comment]
> It's not your fault, you need to rebase with master.
> Or we can just merge it as is

@mgrybyk I have already done that https://github.com/SanthoshBonala/webdriverio/pull/2
do I need to anything else?

**Comment by mgrybyk** [comment]
Ah, got it, moment. 
Nothing from you @SanthoshBonala , thank you!

**Pull request merged**